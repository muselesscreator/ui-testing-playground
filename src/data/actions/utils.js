import { createAction } from '@reduxjs/toolkit';

const createActionFactory = (dataKey) => (actionKey, ...args) => (
  createAction(`${dataKey}/${actionKey}`, ...args)
);

export {
  // eslint-disable-next-line import/prefer-default-export
  createActionFactory,
};
