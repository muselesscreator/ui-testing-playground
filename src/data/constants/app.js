import { getConfig } from '@edx/frontend-platform';

// eslint-disable-next-line import/prefer-default-export
export const routePath = `${getConfig().PUBLIC_PATH}:courseId`;
