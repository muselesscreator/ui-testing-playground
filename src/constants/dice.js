import { StrictDict } from 'utils';

export const diceTypes = StrictDict({
  d4: 'd4',
  d6: 'd6',
  d8: 'd8',
  d10: 'd10',
  d12: 'd12',
  d20: 'd20',
  d100: 'd100',
});

export const diceSides = StrictDict({
  [diceTypes.d4]: 4,
  [diceTypes.d6]: 6,
  [diceTypes.d8]: 8,
  [diceTypes.d10]: 10,
  [diceTypes.d12]: 12,
  [diceTypes.d20]: 20,
  [diceTypes.d100]: 100,
});

const tgcURI = 'https://s3.amazonaws.com/files.thegamecrafter.com';
export const diceURIs = StrictDict({
  [diceTypes.d4]: `${tgcURI}/23ab580e2280cd20f2ef36260eb2b8902132e5f7`,
  [diceTypes.d6]: `${tgcURI}/19337628523391821dd8aa251a925afb611932fe`,
  [diceTypes.d8]: `${tgcURI}/c123717767de732ae400c7cc0e193581ee444987`,
  [diceTypes.d10]: 'https://iambesser.me/DA/images/10/1D10.png',
  [diceTypes.d12]: `${tgcURI}/292a945e3ab648e1e24ffa6c8e130c68a8cfa3ff`,
  [diceTypes.d20]: 'https://i.ya-webdesign.com/images/d20-1-png-3.png',
  [diceTypes.d100]: 'https://cdn01.zipify.com/images/000/182/977/original/2179683_20171023T152757.png',
});
