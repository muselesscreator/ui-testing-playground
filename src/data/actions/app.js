import { StrictDict } from 'utils';
import { createActionFactory } from './utils';

export const dataKey = 'app';
const createAction = createActionFactory(dataKey);

/**
 * Add a new roll string to the rolls list
 * @param {string} roll - new roll string
 */
export const addRoll = createAction('add_roll');

/**
 * Clears all rolls from the rolls list
 */
export const clearRolls = createAction('clear_rolls');

export default StrictDict({
  addRoll,
  clearRolls,
});
