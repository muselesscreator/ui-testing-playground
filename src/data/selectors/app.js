import { createSelector } from 'reselect';

import { StrictDict } from 'utils';

import * as module from './app';

export const appSelector = (state) => state.app;

export const simpleSelectors = {
  rolls: createSelector([module.appSelector], (app) => app.rolls),
  lastRollsByType: createSelector(
    [module.appSelector],
    (app) => app.lastRollsByType,
  ),
};

export const rollSummaries = createSelector(
  [module.simpleSelectors.rolls],
  (rolls) => rolls.map(
    ({ rollIndex, type, value }) => ({
      rollIndex,
      summary: `You rolled ${value} on a ${type}`,
    }),
  ),
);

export const lastRollByType = (state, { type }) => (
  module.simpleSelectors.lastRollsByType(state)[type]
);

export default StrictDict({
  ...simpleSelectors,
  lastRollByType,
  rollSummaries,
});
