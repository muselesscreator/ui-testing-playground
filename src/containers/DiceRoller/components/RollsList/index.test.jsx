import React from 'react';
import { shallow } from 'enzyme';

import { diceTypes } from 'constants/dice';

import selectors from 'data/selectors';
import actions from 'data/actions';

import { RollsList, mapStateToProps, mapDispatchToProps } from '.';

jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    app: {
      rollSummaries: jest.fn((...args) => ({ rolls: args })),
    },
  },
}));
jest.mock('@edx/paragon', () => ({
  Button: () => 'Button',
  Image: () => 'Image',
}));

let el;

describe('RollsList component', () => {
  const props = {
    rolls: [
      {
        summary: `You rolled 2 on a ${diceTypes.d4}`,
        rollIndex: 0,
      },
      {
        summary: `You rolled 5 on a ${diceTypes.d10}`,
        rollIndex: 1,
      },
      {
        summary: `You rolled 18 on a ${diceTypes.d20}`,
        rollIndex: 2,
      },
    ],
    clearRolls: jest.fn().mockName('this.props.clearRolls'),
  };
  test('snapshot', () => {
    expect(shallow(<RollsList {...props} />)).toMatchSnapshot();
  });
  describe('component', () => {
    beforeEach(() => {
      el = shallow(<RollsList {...props} />);
    });
    test('a roll summary for each roll', () => {
      const rolls = el.find('.roll-entry');
      expect(rolls.at(0).text()).toEqual(props.rolls[0].summary);
      expect(rolls.at(1).text()).toEqual(props.rolls[1].summary);
      expect(rolls.at(2).text()).toEqual(props.rolls[2].summary);
      expect(rolls.length).toEqual(3);
    });
  });

  describe('mapStateToProps', () => {
    const testState = { a: 'fake state' };
    let mapped;
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('rolls is loaded from selectors.app.rolls', () => {
      expect(mapped.rolls).toEqual(selectors.app.rollSummaries(testState));
    });
  });

  describe('mapDispatchToProps', () => {
    test('maps actions.app.clearRolls to clearRolls prop', () => {
      expect(mapDispatchToProps.clearRolls).toEqual(actions.app.clearRolls);
    });
  });
});
