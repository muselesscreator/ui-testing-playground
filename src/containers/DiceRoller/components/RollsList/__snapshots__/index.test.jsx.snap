// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RollsList component snapshot 1`] = `
<div
  className="rolls-list"
>
  <div
    className="rolls-container"
  >
    <div
      className="roll-entry"
      key="0"
    >
      You rolled 2 on a d4
    </div>
    <div
      className="roll-entry"
      key="1"
    >
      You rolled 5 on a d10
    </div>
    <div
      className="roll-entry"
      key="2"
    >
      You rolled 18 on a d20
    </div>
  </div>
  <div
    className="action-bar"
  >
    <Button
      className="clear-rolls-btn"
      onClick={[MockFunction this.props.clearRolls]}
    >
      Clear
    </Button>
  </div>
</div>
`;
